class.c

```diff
@@ -83,23 +83,25 @@ RUBY_EXTERN rb_serial_t ruby_vm_global_cvar_state;
 static rb_subclass_entry_t *
 push_subclass_entry_to_list(VALUE super, VALUE klass)
 {
-    rb_subclass_entry_t *entry, *head;
-
-    entry = ZALLOC(rb_subclass_entry_t);
+    rb_subclass_entry_t *entry = ZALLOC(rb_subclass_entry_t);
     entry->klass = klass;
 
-    head = RCLASS_SUBCLASSES(super);
-    if (!head) {
-        head = ZALLOC(rb_subclass_entry_t);
-        RCLASS_SUBCLASSES(super) = head;
-    }
-    entry->next = head->next;
-    entry->prev = head;
+    RB_VM_LOCK_ENTER_NO_BARRIER();
```

Comment:
Is `RB_VM_LOCK_ENTER_NO_BARRIER()` correct or it should be `RB_VM_LOCK_ENTER()`? I'm not clear on the difference, after taking a quick look at the sources.
It seems most places use `RB_VM_LOCK_ENTER()`

Reply:
`RB_VM_LOCK_ENTER` will trigger a barrier which stops other Ractors. For correctness here, we just need to ensure that two Ractors do not modify the same superclass at the same time, so we can use the no barrier variant to ensure that.

However, I suspect that reading the subclass list is not synchronized. So by not having a barrier here we run the risk of race conditions in other Ractors that may read an incorrect list.

Reply:
Ok yeah I'm convinced that `RB_VM_LOCK_ENTER` doesn't actually trigger a barrier. I switched it to that 👍

Reply:
I'm glad I asked, this "barrier" is not exactly trivial or the usual meaning of barrier :)
It would be nice to document what RB_VM_LOCK_ENTER_NO_BARRIER is for/when it should be used in the source code, cc @ko1

Reply:
`RB_VM_LOCK_ENTER` get lock and if another Ractor is acquiring the barrier wait, the ractor join this barrier (wait for the barrier locking of another Ractor is finished, maybe by GC).

`RB_VM_LOCK_ENTER_NO_BARRIER` doesn't join this barrier waiting.